common.h

```diff
@@ -0,0 +1,30 @@
+#include<stdio.h>
+#include<stdint.h>
+//#include<cheri/cheri.h>
+#include <cheriintrin.h>
+#include<stdbool.h>
+
+void inspect_pointer(void *ptr)
+{
+	
+	uint64_t length = cheri_length_get(ptr);
+	uint64_t address = cheri_address_get(ptr);
+	uint64_t base = cheri_base_get(ptr);
+	uint64_t flags = cheri_flags_get(ptr);
+	uint64_t perms = cheri_perms_get(ptr);
+	uint64_t type = cheri_type_get(ptr);	
+	bool tag = cheri_tag_get(ptr);
+	bool valid = cheri_is_valid(ptr);
+
+	uint64_t offset = __builtin_cheri_offset_get(ptr);
+	printf("Address: %04lx, Base: %04lx, End: %04lx Flags: %04lx, Length: %04lx, Offset: %04lx, Perms: %04lx, Type: %04lx, Tag: %d, Valid: %d\n", address, base, base + length, flags, length, offset, perms, type, tag, valid);
+}
+
+// This hsould be a macro or inlined
```

Comment:
Why?

Reply:
I'm only guessing  but the (currently commented out) use of `inspect_pointer` in the **stackscan.c** example indicates we may wish to use `inspect_pointer` whilst traversing and inspecting the stack. Inlining via macro expansion avoids the issue of mutating the stack if the call to `inspect_pointer` is not otherwise inlined by the compiler.

Reply:
Added an answer it the comment: https://github.com/capablevms/cheri-examples/blob/1f8537c5e8e99633fb816022662eb0d728c14d84/include/common.h#L28-L34

Reply:
I think the reasoning is sound, roughly the same as the above comment regarding `inspect_pointer`.
Using the `inline` keyword will still require that you provide a non-inline implementation that manipulates the `csp` value.
Using a macro has the advantage that you're not dependent on the optimisation level selected during compilation.

Out of these two I would still prefer the `inline` keyword version. I'm happy to `resolve` the issue subject to an issue being opened describing the chosen course of action, unless there is any objection.

Reply:
I still didn't understand why it should be a macro or inlined :)

Reply:
Then I would ask do you think this is fine like that?

Reply:
OK, so Aleks and I spoke offline. I now understand the comment, I think. And, assuming I do understand it, I subtly disagree with it :) I agree that `cheri_csp_get` over-approximates the stack, in the sense that you get a pointer that captures more of the stack than you really want. *However* if you were to make this a macro, or inline it, then it's possible that the compiler would decide to move the `asm` statement around, possibly under-approximating the stack -- which would be bad. Now, the compiler probably won't do this, because it tends to be very conservative when it sees `asm` statements, but back when I did a conservative stack scanner in C, I found that having the `asm` statement in a function I "cared" about did lead to subtle bugs.

Long story short, I think the comment should be something more like:

```
// This function returns the current stack pointer.
//
// It might seem that this should be a macro or an inlined function, but then we couldn't be
// entirely sure where the compiler might move the `asm` statement to.
```

That suggests that we should probably mark this function as "never inline".

An alternative approach might be to use compiler barriers to stop reordering. I haven't thought much about those in the context of C, and frankly, it's probably overkill to do so here, but it might be an interesting thing for you to have a look at one day!

Reply:
I like the `never inline` approach, it makes the semantics consistent with the minimum level of effort. The  [optnone](https://releases.llvm.org/3.8.0/tools/clang/docs/AttributeReference.html#optnone-clang-optnone) attribute may be of interest in similar contexts.

Reply:
I think I have a new better solution to this problem: 
https://github.com/capablevms/cheri-examples/blob/69dc074e5bd35209b82fce2bb2db478dcc62d368/include/common.h#L30-L36